### DISCUSSION TITLE
Inference problems in PYNQ-Z1

### DISCUSSION THREAD

**Discussion by ISPRrj** [Xilinx/finn#849] [Q&A]
Hi everyone!

I am trying to implement the Lenet5 network on the PYNQ-Z1 board. For that purpose I have created the network using brevitas and I have obtained the following accuracy after training (about 55%).

![brevitasLenet](https://github.com/Xilinx/finn/assets/124183109/99ed9dad-a351-49cb-8b4c-561652e0f7e7)

![Brevitas accuracy](https://github.com/Xilinx/finn/assets/124183109/51745332-dc5f-4de9-a5db-7dff876e077f)

I have followed all the steps of the end to end flow FINN and even all the intermediate verifications (including the emulation using PyVerilator) and everything seems to work correctly. But at the moment of performing the inference with the PYNQ board, the accuracy obtained is only 8%, far below the 55% obtained with brevitas.....

Any help on why this could be happening or how to debug to see the problem?

**Comment by auphelia** [comment]
Hi @ISPRrj ,

Without having more context about how your network looks like in ONNX and which FINN version you're using, I can only speculate. 

What I could imagine is that the last QuantLinear layer causes a floating-point multiplication at the end of the ONNX graph (from the scaling factor) that cannot be streamlined away. This last layer can then not be implemented using an HLS layer in FINN and will be excluded from the hardware generation, you will need to apply the scaling factor then manually in the driver. You can check this theory by having a look at the graphs before and after HLS conversion, to see which layers were converted and which weren't.

**Comment by ISPRrj** [comment]
Thank you very much for your response @auphelia!

I'm using FINN v0.9. At the beginning, as you say, the end nodes referring to the last layer (Quantlinear) could not be implemented using an HLS layer. But I removed the bias of the last layer and made the appropriate transformations to finally have only HLS layers (except the initial Transpose node).

This is what the ONNX chart looks like after **pre and post processing**.

![lenet_quantized_pre_post onnx](https://github.com/Xilinx/finn/assets/124183109/dcc3ad22-ee75-4e5e-816b-1c36cbc258b1)

I then perform the transformations:

model = ModelWrapper("lenet_quantized_pre_post.onnx")
model = model.transform(MoveScalarLinearPastInvariants())
model = model.transform(Streamline())
model = model.transform(LowerConvsToMatMul())
model = model.transform(MakeMaxPoolNHWC())
model = model.transform(absorb.AbsorbTransposeIntoMultiThreshold())


model = model.transform(MakeMaxPoolNHWC())
model = model.transform(absorb.AbsorbConsecutiveTransposes())

model = model.transform(Streamline())

model = model.transform(absorb.AbsorbScalarMulAddIntoTopK())
model = model.transform(InferDataLayouts())
model = model.transform(RemoveUnusedTensors())
model.save("lenet_quantized_streamlined.onnx")

model = ModelWrapper("lenet_quantized_streamlined.onnx")

This is what it looks like just **before the conversion to HLS layers**.

![lenet_quantized_streamlined onnx](https://github.com/Xilinx/finn/assets/124183109/f0e69f9e-00ec-4493-8d86-377fbad10818)

mem_mode = "decoupled"

model = model.transform(to_hls.InferBinaryMatrixVectorActivation(mem_mode))
model = model.transform(to_hls.InferQuantizedMatrixVectorActivation(mem_mode))


model = model.transform(to_hls.InferLabelSelectLayer())
model = model.transform(to_hls.InferThresholdingLayer())
model = model.transform(GiveUniqueNodeNames())


model = model.transform(to_hls.InferThresholdingLayer())
model = model.transform(to_hls.InferConvInpGen())
model = model.transform(to_hls.InferStreamingMaxPool())

model = model.transform(RemoveCNVtoFCFlatten())

model = model.transform(absorb.AbsorbConsecutiveTransposes())

model = model.transform(InferDataLayouts())

model.save("lenet_hls_layers.onnx")

And finally this is the ONNX graph after **the implementation of the HLS layers**.

![flowers_hls_layers onnx](https://github.com/Xilinx/finn/assets/124183109/45b8efaf-37e1-483c-8466-9660c4214de1)

One thing I would like to comment on is when I compare the brevitas and FINN inference I note that there are 2 cases out of 100 that do not match ( bit it still does not explain the move from 55% accuracy (brevitas) to 8% (PYNQ). This just before the conversion to HLS layers.

I alsooget the message:

_FINN execution:   0%|                                                                                                                          | 0/100 [00:00<?, ?it/s]/home/jpiera/finn/deps/qonnx/src/qonnx/util/basic.py:300: UserWarning: The values of tensor MultiThreshold_0_out0 can't be represented with the set datatype annotation (INT8), they will be rounded to match the datatype annotation.
  warnings.warn(
ok 98 nok 2: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:20<00:00,  4.94it/s]_

Continuing with the verifications, the simulation using cppsim and the emulation using PyVerilator node by node and stitched IP, were successful.

I'm not sure how to find the problem...

**Comment by auphelia** [comment]
Hi @ISPRrj ,

Yes, I see that the last scale factor is probably merged into the TopK node. Which Vivado version are you using?

Also just to understand your problem better:
- Brevitas/PyTorch works correctly (accuracy is 55% on which dataset?)
- Exported QONNX gives you the same result for a selected input compared to Brevitas/PyTorch or are you also running the verification with the full dataset and you observe 55%?
- After conversion to HLS layers, cppsim and node-by-node rtlsim give the same result as before. Also here on which data? An example input, a random input tensor or on the whole test dataset?
- Stitched IP rtlsim gives the same result as the steps before, with which test input?
- Only the bitfile running on the Pynq-Z1 board behaves differently. Also for the test image or for the test dataset?

As a remark, the warning is only affecting the node-by-node execution in FINN. By default the environmental variable `SANITIZE_QUANT_TENSORS` is set to 1 (can be set to 0). When this env var is set, it gets called before and after the execution of a node, as you can see [here](https://github.com/fastmachinelearning/qonnx/blob/main/src/qonnx/core/onnx_exec.py#L178C48-L178C48). If the execution introduces very small rounding errors, the FINN execution automatically "sanitizes" the values to make sure that the behavior would be like in the generated hardware. You can check out the function [here] (https://github.com/fastmachinelearning/qonnx/blob/main/src/qonnx/util/basic.py#L267).

In the PyTorch description you sent in the beginning it looks like you're quantizing to 2 bit but the warning you've shared says 8 bit. How is your network quantized?

**Reply by ISPRrj** [reply]
Hi @auphelia !

- **Which Vivado version are you using?**
 2022.2

- **Brevitas/PyTorch works correctly (accuracy is 55% on which dataset?)**
I'm using this dataset "https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz"

- **Exported QONNX gives you the same result for a selected input compared to Brevitas/PyTorch or are you also running the verification with the full dataset and you observe 55%?**
I'm using the exported FINN-ONNX  for several (100) selected inputs and I obtain the same result compared to Brevitas.

- **After conversion to HLS layers, cppsim and node-by-node rtlsim give the same result as before. Also here on which data? An example input, a random input tensor or on the whole test dataset?**
I'm comparing 100 example inputs and give the same result as before.
- **Stitched IP rtlsim gives the same result as the steps before, with which test input?**
In this one I was wrong, it turns out that the stitched IP emulation always returns the same output regardless of the input value.

- **Only the bitfile running on the Pynq-Z1 board behaves differently. Also for the test image or for the test dataset?**
It behaves differently for the test image and also for the test dataset (compared to Brevitas).
- **In the PyTorch description you sent in the beginning it looks like you're quantizing to 2 bit but the warning you've shared says 8 bit. How is your network quantized?**
As you say I am using a 2-bit quantization for the weights and activations. But I think that the 8 bit warning is given by the QuantIdentity that I define at the beginning of the class whith a bit width of 8.

**Comment by auphelia** [comment]
@ISPRrj ,

It might be a good idea to turn this into an issue using the corresponding template.

**Reply by auphelia** [reply]
Hi @ISPRrj ,

After reading your latest reply, if you're using 2022.2, could you please switch to using `dev`? And I would recommend to use the stitched IP rtlsim for debugging, when setting the metadata property "rtlsim_trace" before calling the execution function, a wavesim will be generated that you can inspect more closely. 
E.g. `model.set_metadata_prop("rtlsim_trace", "model_trace.vcd")`
Additionally, please open an issue for this and provide the information that is requested through the template.

**Reply by auphelia** [reply]
I am linking the opened issue here: https://github.com/Xilinx/finn/issues/855
